node/test/util/rm.js

```diff
@@ -0,0 +1,396 @@
+            },
+            "rm a file from an open submodule": {
+                initial: "sub=S:C2-1 x/y/a=foo,x/y/z=foo;Bmaster=2|" +
+                    "x=S:C3-1 d=Ssub:2;" +
+                    "Bmaster=3;Od",
+                path: "d/x/y/z",
+                output: {
+                    isRootLevel: false,
+                    pathFromRoot: "d",
+                    pathInSubmodule: "x/y/z"
+                }
+            },
+            "rm a nonexistent dir from an open submodule": {
+                initial: "sub=S:C2-1 x/y/a=foo,x/y/z=foo;Bmaster=2|" +
+                    "x=S:C3-1 d=Ssub:2;" +
+                    "Bmaster=3;Od",
+                path: "d/q",
+                fails: true
+            },
+            "rm a file an open submodule via dir, no --recursive": {
+                initial: "sub=S:C2-1 x/y/a=foo,x/y/z=foo;Bmaster=2|" +
+                    "x=S:C3-1 d=Ssub:2;" +
+                    "Bmaster=3;Od",
+                path: "d/x/y",
+                recursive: false,
+                fails: true
+            },            
+            "rm a file from an open submodule via dir": {
+                initial: "sub=S:C2-1 x/y/a=foo,x/y/z=foo;Bmaster=2|" +
+                    "x=S:C3-1 d=Ssub:2;" +
+                    "Bmaster=3;Od",
+                path: "d/x/y",
+                recursive: true,
+                output: {
+                    isRootLevel: false,
+                    pathFromRoot: "d",
+                    pathInSubmodule: "x/y"
+                }
+            },            
+        };
+        Object.keys(cases).forEach(caseName => {
+            const c = cases[caseName];
+            it(caseName, co.wrap(function *() {
+                const getRemovalSpec = co.wrap(function *(repos) {
+                    const repo = repos.x;
+                    const index = yield repo.index();
+                    const res = yield Rm.getRemovalSpec(repo, index, c.path,
+                                                         c.recursive || false);
+                    assert(deeper(c.output, res));
+                });
+                yield RepoASTTestUtil.testMultiRepoManipulator(c.initial,
+                                                               c.initial,
+                                                               getRemovalSpec,
+                                                               c.fails);
+            }));
+        });
+    });
+
+    describe("rmPaths", function () {
+        const cases = {
+            // Cases 3 and 4
+            "rm everything from empty repo": {
+                initial: "x=S:I README.md",
+                paths: [""],
+                fails: true,
+                expect: "x=S:I README.md",
+            },
+            "rm a clean submodule by name": {
+                initial: "x=S:C2-1 x/y/z=foo;W x/q/z=bar;Bmaster=2",
+                paths: ["x/y/z"],
+                expect: "x=S:C2-1 x/y/z=foo;W x/q/z=bar;I x/y/z;Bmaster=2",
```

Comment:
Another simplification: the repo type "E" (for "expected") starts out as the case used initially, so you could simplify "expect" to: `x=E:I x/y/z` -- just specifying what's changed.